File: benchmark/cases/k_nucleotide/k_nucleotide.py
import os
import time

base = "GGTATTGAGCACTGGCAATTGACGTCAGGTATCCGAATTGCACGTTAGCATGCATGCATGCACGT"
s = base * int(os.environ["BENCH_SCALE"])
h = {}
t0 = time.perf_counter_ns()
for ch in s:
    c = ord(ch)
    h[c] = h.get(c, 0) + 1
t1 = time.perf_counter_ns()
print(
    "knuc: A =", h[65], "C =", h[67], "G =", h[71], "T =", h[84]
)
print("__bench_ns:", t1 - t0)

File: benchmark/cases/k_nucleotide/k_nucleotide.rs
use std::collections::HashMap;
use std::time::Instant;

fn main() {
    let base = "GGTATTGAGCACTGGCAATTGACGTCAGGTATCCGAATTGCACGTTAGCATGCATGCATGCACGT";
    let scale: usize = std::env::var("BENCH_SCALE").unwrap().parse().unwrap();
    let s = base.repeat(scale);
    let mut h: HashMap<i64, i64> = HashMap::new();
    let t0 = Instant::now();
    for ch in s.chars() {
        let c = ch as i64;
        *h.entry(c).or_insert(0) += 1;
    }
    let ns = t0.elapsed().as_nanos();
    println!(
        "knuc: A = {} C = {} G = {} T = {}",
        h[&65], h[&67], h[&71], h[&84]
    );
    println!("__bench_ns: {}", ns);
}
